Delivery Service Active Flag Rework

ocket8888 opened the pull request:
In accordance with the blueprint from #4654, this changes the `active` property of Delivery Services from a simple boolean to a three state system, where the three states are:

- ACTIVE - The DS is routed and cache servers will generate configuration for it. The equivalent of `true` today.
- PRIMED - The DS is **not** routed, but cache servers will still generate configuration for it. The equivalent of `false` today.
- INACTIVE - The DS is not routed **and** cache servers will **not** generate configuration for it. Has no equivalent today.

Additionally, because I was making a new DS model already anyway since that's a breaking change, I made some things that are never allowed to be null or undefined into non-reference values (e.g. XMLID is a `string` not a `*string` etc.) and made `lastUpdated` format in RFC3339 in JSON responses.

The new model now totally drops `LongDesc1` and `LongDesc2`, which were deprecated in the now-itself-deprecated APIv3 and unavailable for direct use in APIv4 - as an interesting consequence of this and how DSRs are stored internally, **it is no longer possible to make a DSR that will set a DS's `longDesc1` and/or `longDesc2` to anything but `null` _even using APIv3_**. They are shown as `null` in responses, but are otherwise totally ignored by Traffic Ops. DS `longDesc1` and `longDesc2` fields can still be manipulated directly and viewed as normal in APIv3.

I also made some changes to the testing utilities, because they weren't registering themselves as helpers so it was difficult to find actual failures.

Fixes #3746

This PR includes documentation, TO changes, and database migrations, but does not actually make the new state work in `t3c`, nor does it update Traffic Portal with the ability to deal with it. These efforts are left to a later date.

<hr/>

## Which Traffic Control components are affected by this PR?
- Documentation
- Traffic Ops Client (Go)
- Traffic Ops

## What is the best way to verify this PR?
I've updated the tests, and really there's only one new DS permutation, which has a test case added, so there's not really a whole lot more to do than just run the tests. You can replicate them by using APIv5 to manipulate DSes and DSRs if you like, just make sure everything looks okay.

## If this is a bugfix, which Traffic Control versions contained the bug?
- 7.x
- 6.x

## PR submission checklist
- [x] This PR has tests
- [x] This PR has documentation
- [x] This PR has a CHANGELOG.md entry
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY**

rimashah25 commented:
Tested the API tests locally, all work (v3/v4/v5) fine. Yet to test DS DSRs creation with API v5 curl cmds.

rimashah25 commented on traffic_ops/v5-client/deliveryservice_requests.go:
```diff
@@ -58,31 +58,31 @@ func (to *Session) CreateDeliveryServiceRequest(dsr tc.DeliveryServiceRequestV4,
 	}
 
-	if ds.TypeID == nil && ds.Type.String() != "" {
+	if ds.TypeID <= 0 && ds.Type != nil && *ds.Type != "" {
```
What is the difference between type and typeID. As I understand it is the name and ID of a given type, so why check for int, string and pointer?

ocket8888 replied:
The Type is known to humans by its name, so this block was meant to populate the ID if it wasn't given by using the Type's Name (if that *is* given, otherwise it's an error). Because `ds.Type` is a pointer, it can be `nil`, so its value must not be accessed in that case, or a bad error response will be returned to the user. You also can't look up a Type with a blank Name, because Type Names can't be blank. The old code called `.String()`, which would segfault if the `ds.Type` reference was `nil`, so I added the `nil` check to avoid that. `TypeID` is no longer a pointer, so instead of `nil` I checked for the "zero" value of the property's datatype (and also invalid values i.e. negative ones).

rimashah25 replied:
Ok, Thanks for the explanation. You can resolve this comment.

rimashah25 commented on traffic_ops/testing/api/v5/deliveryservice_request_comments_test.go:
```diff
@@ -178,7 +178,9 @@ func CreateTestDeliveryServiceRequestComments(t *testing.T) {
 		resp, _, err := TOSession.GetDeliveryServiceRequests(opts)
 		assert.NoError(t, err, "Cannot get Delivery Service Request by XMLID '%s': %v - alerts: %+v", comment.XMLID, err, resp.Alerts)
-		assert.Equal(t, len(resp.Response), 1, "Found %d Delivery Service request by XMLID '%s, expected exactly one", len(resp.Response), comment.XMLID)
+		if !assert.Equal(t, len(resp.Response), 1, "Found %d Delivery Service request by XMLID '%s, expected exactly one", len(resp.Response), comment.XMLID) {
```
Why continue and not error out?

ocket8888 replied:
Because although creation of the currently processed comment has failed (or at best succeeded wrongly), other comments can still be created, which can help narrow down what the problem was and/or find other problems by continuing to run other tests.

rimashah25 commented on traffic_ops/testing/api/v4/tc-fixtures.json:
```diff
@@ -907,7 +907,7 @@
             "regionalGeoBlocking": false,
             "remapText": null,
-            "routingName": "cdn",
+            "routingName": "ccr-ds1",
```
Why set it to an invalid value? What scenario are we trying to test?

ocket8888 replied:
That value is not invalid, what it was set to *before* was actually hiding an error. Changing this was part of the fix for #7094.

rimashah25 replied:
oh.. ok

ocket8888 replied:
No, it seems we don't

rimashah25 commented on traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go:
```diff
@@ -1023,6 +1027,7 @@ func scanDSInfoRow(row *sql.Row) (DSInfo, bool, error) {
 		}
 		return DSInfo{}, false, fmt.Errorf("querying delivery service server ds info: %v", err)
 	}
+	di.Active = active == tc.DSActiveStateActive
```
I am so lost here, why is a boolean being assigned a string and then use a comparison operator?

rimashah25 commented on traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go:
```diff
@@ -1023,6 +1027,7 @@ func scanDSInfoRow(row *sql.Row) (DSInfo, bool, error) {
 		}
 		return DSInfo{}, false, fmt.Errorf("querying delivery service server ds info: %v", err)
 	}
+	di.Active = active == tc.DSActiveStateActive
```
I am so lost here, why is a boolean being assigned a string and then use a comparison operator?

ocket8888 replied:
A boolean is not being assigned to a string. A boolean is being assigned to the outcome of comparing two strings. A Delivery Service where `"active": true` in the old data model is equivalent to `"active": "ACTIVE"` in the new data model. So if `active` is exactly `tc.DSActiveStateActive`, then `di.Active` should be `true`. If it's anything else, `di.Active` should be false. Which is what this line does.

rimashah25 commented on traffic_ops/app/db/migrations/2022092107561215_ds_explicit_mso.down.sql:
```diff
@@ -0,0 +1,20 @@
+ * the License.
+ */
+
+ALTER TABLE public.deliveryservice
```
The up.sql file has updates on `deliveryservice_request` table but I see nothing to revert that in down file?

ocket8888 replied:
Those changes don't need to be reverted. Before the change, `multiSiteOrigin` can be one of three values: `true`, `false`, or `null`. The `up` file changes all `null` values to `false` (because `null` and `false` are treated exactly the same, so there was no reason to allow both). The `up` file needed to do that because after it runs, `null` will no longer be allowed. But after `down` runs, `false` is still allowed - and also it's treated the same as `null` so there's no behavioral impact, and finally you also can't know which `false`s used to be `null`s and which were `false` in the first place, so the information needed to restore that state doesn't exist anyway.

ocket8888 commented on traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:
```diff
@@ -138,7 +162,7 @@ func CreateV30(w http.ResponseWriter, r *http.Request) {
 	ds := tc.DeliveryServiceV30{}
 	if err := json.NewDecoder(r.Body).Decode(&ds); err != nil {
-		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("decoding: "+err.Error()), nil)
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, fmt.Errorf("decoding: %w", err), nil)
```
> Why the change?

Because it preserves the identity of the underlying error that way. Here's [a playground link to an example program that shows what that means](https://go.dev/play/p/GOkCOgci1lh). It accomplishes the same thing as `errors.New(someString+err.Error())` but without destroying any information.

> ...shouldn't we refactor others in the code base too?

In my opinion, yes, and I always request people do it this way in my reviews. I only changed it in the files I was already editing, though, because this is done untold thousands of times throughout ATC. Nobody would ever review that PR. So I just make smaller changes to files as I work on them.

rimashah25 replied:
Ok, I like that you are taking care of it in each of your PR but I feel that is a bit of scope creep. Because this is refactor(necessary) but I am not a 100% sure in a PR this big it should have been included. I
f you want, create an issue to tackle this as part of tech-debt and we can take care of it at one go.

ocket8888 replied:
took it out

ocket8888 commented on traffic_ops/traffic_ops_golang/deliveryservice/safe.go:
```diff
@@ -144,19 +141,28 @@ func UpdateSafe(w http.ResponseWriter, r *http.Request) {
 			}
-			api.WriteRespAlertObj(w, r, tc.SuccessLevel, alertMsg, []tc.DeliveryServiceV30{ds.DowngradeToV31().DeliveryServiceV30})
+			api.WriteRespAlertObj(w, r, tc.SuccessLevel, alertMsg, []tc.DeliveryServiceV30{ret.DeliveryServiceV30})
 		case 2:
```
Yes, though when I first made these changes we had not, yet.

ocket8888 commented on traffic_ops/testing/api/v5/deliveryservices_keys_test.go:
```diff
@@ -67,30 +67,29 @@ func createBlankCDN(cdnName string, t *testing.T) tc.CDN {
 }
 
-func cleanUp(t *testing.T, ds tc.DeliveryServiceV4, oldCDNID int, newCDNID int, sslKeyVersions []string) {
-	if ds.ID == nil || ds.XMLID == nil {
```
Because XMLID can no longer be `nil`. If that check was still there, the tests wouldn't compile.

ocket8888 commented:
This PR depends on #7156

ocket8888 commented:
I need to change the migration file names because new migrations have been added since this PR opened. I'm in a meeting atm, I'll do it this afternoon.

rimashah25 commented:
Tested and reviewed the following:
1. Able to create DS with 3 states (Active, Primed, Inactive) using APIv5
2. Able to create DS with 2 states (as expected) (Active=true, Primed=false) using APIv3, APIv4, APIv4.1
3. See the timestamp format update in APIv5 ONLY
4. Could create DS and DSR with longDesc1 and longDesc2 in the JSON payload with all 3 API versions 
5. #7094 didn't seem like it was fixed. I am able to change the routingName even when the value is set to cdn 
6. GET receives the right DS structure (3.0 and 4.0 doesn't show `regional` field but 4.1 and 5.0 shows it)
7. Update to testing utilities makes sense to implement.
8. Variable name changes are tested in all three versions of API tests
9. DB migration files need timestamp change.
10. Remaining to test the following API: `/cdns/{{CDN Name}}/configs/monitoring`, `/cdns/{{CDN Name}}/snapshot/new`, `/deliveryservices/{{ID}}/safe`, `/servers/{{ID}}/deliveryservices`, `/snapshot`, `/users/{{ID}}/deliveryservice`

ocket8888 commented:
It looks like I accidentally removed `regional` from the docs when I rebased them after it was added, so I'll fix that and the migration timestamps. Other than that:

>     4. Could create DS and DSR with longDesc1 and longDesc2 in the JSON payload with all 3 API versions. Per the description in the PR, I shouldn't be able to do so.

I think maybe I should clarify. You should be able to create a DS and DSR with any extra fields specified that you want. You could set `"foo": "bar"` and it shouldn't break anything. But it just shouldn't be stored. The same thing should happen with `longDesc1` and `longDesc2` in APIv5. Requests containing those should succeed, but no matter what you set them to, they should be `null` when you GET it in APIv3. Also, because DSRs are always stored at the latest API version, you should be able to create a DSR that sets a DS's `longDesc1`/`longDesc2` to anything you want, but even if you do that in APIv3 it should be null when you GET it afterward.

Also, the legacy structures for DSes use the `omitempty` directive in the `json` portion of their struct tags for those fields, so when `longDesc1`/`longDesc2` is/are `NULL` in the database, it/they are omitted from the response, so in that case it looks the same as if I had removed them from that version too (but I didn't I swear). Basically those will always be missing from DSRs in APIv3 but that's the same as if you used that version to set them explicitly to null before these changes.

That said, it looks like DSRs are always showing APIv5 representations in responses to POST requests on this branch - which doesn't fail tests because I guess the APIv3 client doesn't decode/return the `response` to those requests. Which I can't add because it'd be a breaking change, so I'll just need to fix and test manually, since a test for that can't really be written easily.

>     5. [Routing Name can be changed to default #7094](https://github.com/apache/trafficcontrol/issues/7094) didn't seem like it was fixed. I am able to change the routingName even when the value is set to cdn

That issue doesn't really exist. I think when I checked for it on master I forgot to generate SSL Keys for the delivery service first. I'll close that.

>     11. `/users/{{ID}}/deliveryservices` cannot curl this API with any versions

This endpoint existed in API versions 1 and 2, which existed when that blueprint was written but no longer do. So it's fine to be unreachable, since it would've been before anyway.

rimashah25 commented:
Today, if I give `longDesc1` and `longDesc2` in 4.0 this is the error I see:
`curl -k -X POST -H "Content-Type: application/json" --cookie "mojolicious=eyJhdXRoX2RhdGEiOiJhZG1pbiIsImV4cGlyZXMiOjE2Njc1OTc4MjIsImJ5IjoidHJhZmZpY2NvbnRyb2wtZ28tdG9jb29raWUifQ--f85d59e0865cbcf9bac01d3fc42b5e12e7c8f8bf" --data @ds_v4.json "https://localhost:8443/api/4.0/deliveryservices"
{"alerts":[{"text":"the longDesc1 and longDesc2 fields are no longer supported in API 4.0 onwards","level":"error"`

So, why is the same strategy not followed for 4.0 and 5.0 in this PR?

ocket8888 commented on docs/source/api/v5/deliveryservices_id.rst:
```diff
@@ -400,20 +404,25 @@ Response Structure
 			"text": "ds was deleted.",
 			"level": "success"
+		},
+		{
+			"text": "Perform a CDN snapshot then queue updates on all servers in the cdn for the changes to take effect.",
+			"level": "info"
```
Possibly, but that should be a separate effort, since this was added in a different PR.

ocket8888 commented on docs/source/api/v5/deliveryservice_requests.rst:
```diff
@@ -663,59 +504,315 @@ The response is a full representation of the edited :term:`Delivery Service Requ
 			"innerHeaderRewrite": null,
+			"ipv6RoutingEnabled": true,
 			"lastHeaderRewrite": null,
+			"lastUpdated": "2020-02-13T16:43:54Z",
```
I'm not sure what you mean; [that *is* RFC3339 format](https://go.dev/play/p/QZuY0kwsep8).

rimashah25 replied:
So, I do see lastUpdated having some decimal values before `Z`, hence I asked. Look at L708 asn example.

rimashah25 replied:
Copy! Looked like two different format, hence the confusion but thanks for the explanation.

ocket8888 commented on docs/source/api/v5/deliveryservice_requests.rst:
```diff
@@ -214,18 +303,19 @@ The request must be a well-formed representation of a :term:`Delivery Service Re
 		"status": "draft",
 		"requested": {
-			"active": false,
+			"active": "INACTIVE",
```
I'm sorry, I'm not sure I know what you mean here.

ocket8888 commented on traffic_ops/testing/api/v3/tc-fixtures.json:
```diff
@@ -858,7 +858,7 @@
             "regexRemap": null,
             "regionalGeoBlocking": false,
             "remapText": null,
-            "routingName": "cdn",
```
I don't think so. I really hope the tests don't start failing when I push that up

rimashah25 replied:
Okie..

ocket8888 commented on traffic_ops/testing/api/v5/tc-fixtures.json:
```diff
@@ -341,7 +341,7 @@
         {
```
it would seem not. The PR that added that field didn't add them to the fixtures and the tests all pass fine. I believe that's abusing the fact that if a field isn't found when unmarshaling, Go will just leave it as the "zero" value of its type. It's not proper IMO, but it's also outside the scope of my PR to fix

ocket8888 replied:
#7189

rimashah25 commented on traffic_ops/app/db/migrations/2022110908494015_ds_active_flag.up.sql:
```diff
@@ -0,0 +1,68 @@
+	deliveryservice ? 'active'
+	AND
+	(deliveryservice -> 'active')::boolean IS TRUE;
+UPDATE public.deliveryservice_request
```
Don't we need a new line after each update?

rimashah25 reviewed (commented):
A few minor comments but overall the PR LGTM from functionality and integration tests perspective.

rimashah25 commented:
All comments have been resolved.

rimashah25 commented:
LGTM, ready for a merge.

ocket8888 commented:
Done.

rimashah25 commented on traffic_ops/testing/api/utils/utils.go:
```diff
@@ -119,7 +130,7 @@ type V3TestData struct {
 
 // V3TestDataT represents the data needed for testing the v3 api endpoints.
 type V3TestDataT[B any] struct {
-	EndpointId     func() int
+	EndpointID     func() int
```
I agree it creates a bigger diff and had removed. But since I had reviewed it multiple times already, I asked him to add it to give continuity to my review.

zrhoffman replied:
Even if you reviewed it and keeping it lends continuity to your review, that doesn't make it have to do with the feature added in the PR itself. It does not belong in this PR.

rimashah25 replied:
I understand but removing at this stage will be a bigger effort than necessary. So, IMO, I would let it be in hopes that this is not repeated in future.

zrhoffman replied:
Removing it at this stage is not difficult, it's a single *refactor* command per field in an IDE

zrhoffman replied:
It's not just my personal opinion, it's the guidelines that our community voted on. Under [*One pull request per feature*](https://github.com/apache/trafficcontrol/blob/master/CONTRIBUTING.md#one-pull-request-per-feature):

> Like big functions, big pull requests are just hard to review. Make each pull request as small as possible.

There's a reason we have that guideline, and it's for cases like these.

rimashah25 replied:
I have nothing more to add since I get where you are coming from..  each developer tries and does scope creep.. some are small and we let it go.. some are big (like the current one) which is hard to review..  I have made peace with a big diff on the PR and I'm ok to let it (scope creep) go this time. I would be very grateful if you let it too.

zrhoffman replied:
> I would be very grateful if you let it too.

Nope

zrhoffman replied:
It's not just a little change. Reverting it brings *Files changed* for this PR down from 160 to 100.

zrhoffman replied:
Correction: Brings it down from 160 to 60

ocket8888 replied:
Removing is very simple (if somewhat annoying because I already did that and then un-did it) but I don't think that work fully captures what Rima means by "it would be more work at this point to remove it". It doesn't belong in the PR, that's true, but it's already been reviewed as a part of this PR. It was a _simple_ change, but not a _small_ one. So to pull that out into a separate PR is duplicating work already spent on reviewing those changes here, which isn't insignificant. Having already volunteered yourself for the task, it's clear you're prepared to take on that tedium, but consider that the time it took Rima to review it already will have been wasted, which is why she asked me to put it back in after I took it out.

At this point I've put it in, taken it out, and put it back in again. I truly do not care if it goes in this PR, but if I change it again I want to be assured it'll be the last time. Perhaps get a third reviewer to break the tie. Or don't if you can come to a consensus without that. But I just need to know what I have to do.

zrhoffman replied:
I'm fine with you keeping the 100-file field rename in #7099. Separately, I've removed those so I could see what the API tests diff looks like without that part, and it looks like #7099 is missing test coverage for the `"PRIMED"` DeliveryServiceActiveState. #7099 is also missing coverage for an invalid DeliveryServiceActiveState being POSTed or PUT.

ocket8888 replied:
> ...it looks like https://github.com/apache/trafficcontrol/pull/7099 is missing test coverage for the "PRIMED" DeliveryServiceActiveState.

I see two Delivery Services in the fixtures data that use that active state, so it seems to be covered to me. It's also worth noting that all Delivery Services with `active: false` in APIv3 and APIv4 tests are implicitly using "PRIMED"; technically the new state is "INACTIVE", not "PRIMED".

> ... missing coverage for an invalid DeliveryServiceActiveState being POSTed or PUT.

That's true. Ideally it should have both.

rimashah25 replied:
Thank you @zrhoffman for being ok with a big scope creep in this PR.

rimashah25 commented on traffic_ops/testing/api/utils/utils.go:
```diff
@@ -129,7 +140,7 @@ type V3TestDataT[B any] struct {
 
 // V4TestData represents the data needed for testing the v4 api endpoints.
 type V4TestData struct {
-	EndpointId    func() int
+	EndpointID    func() int
```
Same as above comment

ocket8888 replied:
I'm going to resolve all of these that relate to the conversation on the above review comment.

rimashah25 commented on traffic_ops/testing/api/utils/utils.go:
```diff
@@ -139,7 +150,7 @@ type V4TestData struct {
 
 // V5TestData represents the data needed for testing the v5 api endpoints.
 type V5TestData struct {
-	EndpointId    func() int
+	EndpointID    func() int
```
Same as above comment.

zrhoffman commented on lib/go-tc/copy.go:
```diff
@@ -0,0 +1,112 @@
+package tc
```
That's true. What you did is definitely an improvement already, so maybe someone can move it to go-util in a future PR.

ocket8888 commented on traffic_ops/traffic_ops_golang/crconfig/deliveryservice.go:
```diff
@@ -145,7 +145,7 @@ FROM deliveryservice AS d
 LEFT OUTER JOIN profile AS p ON p.id = d.profile
 WHERE d.cdn_id = (select id FROM cdn WHERE name = $1)
-AND d.active = true
+AND d.active = 'ACTIVE'
```
The point of using a constant is to avoid misspellings and so you can change it once to change it everywhere - but this being an enum type in the database as well as a Go constant sort of escapes both of those use cases. Because it's an enum it's not possible to insert a misspelling; that would break the API and fail tests etc. And also if the string used to represent "Active" ever changes, it's not possible to change it in only one place, since it would require a database migration. Plus, using string concatenation for building SQL queries - besides always making me a little nervous inherently - means that knowing what you're doing is safe depends on knowing the exact string value of the constant, so an abstraction could be detrimental to that.

For all of those reasons I chose deliberately not to use the Go constant in queries. If you still want me to change it I will, I just wanted to explain my reasoning.

zrhoffman replied:
> And also if the string used to represent "Active" ever changes, it's not possible to change it in only one place, since it would require a database migration

Changing all of the Go occurrences in one place would still be possible

> Plus, using string concatenation for building SQL queries - besides always making me a little nervous inherently - means that knowing what you're doing is safe depends on knowing the exact string value of the constant, so an abstraction could be detrimental to that.

3 of the 6 queries I commented on are queries that already use concatenation, so concatenation should work there, as well as the 2 query fragments.

For the 3 queries that don't use concatenation, passing `tc.DSActiveStateActive` as a parameter to the prepared statement would work (and, unlike the other cases, would not require a typecast).

That said, if you feel that it's more maintainable as-is, I'm fine with you keeping it like that.

ocket8888 replied:
> Changing all of the Go occurrences in one place would still be possible

Sure, but doing that yields a fundamentally broken piece of software, so idk how useful that is.

> 3 of the 6 queries I commented on are queries that already use concatenation, so concatenation should work there, as well as the 2 query fragments.

Without looking at those concatenations, I would say I probably disagree with their use as well, since I know such instances exist.

> For the 3 queries that don't use concatenation, passing tc.DSActiveStateActive as a parameter to the prepared statement would work (and, unlike the other cases, would not require a typecast).
>
> That said, if you feel that it's more maintainable as-is, I'm fine with you keeping it like that.

I think that's actually probably the best course, is to pass it in for statement preparation. That seems more proper than string concatenation and makes the connection between the Go constant and the database enum even more clear. So that's what I'll do.

zrhoffman reviewed (approved):
Looks good.
